File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/config/builder/EngineImpl.kt
/*
 * Copyright 2017-2020 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.docs.config.builder

// tag::class[]
internal class EngineImpl(manufacturer: String, cylinders: Int, crankShaft: CrankShaft, sparkPlug: SparkPlug) : Engine {
    override var cylinders: Int = 0
    private val manufacturer: String
    private val crankShaft: CrankShaft
    private val sparkPlug: SparkPlug

    init {
        this.crankShaft = crankShaft
        this.cylinders = cylinders
        this.manufacturer = manufacturer
        this.sparkPlug = sparkPlug
    }

    override fun start(): String {
        return "$manufacturer Engine Starting V$cylinders [rodLength=${crankShaft.rodLength ?: 6.0}, sparkPlug=$sparkPlug]"
    }

    class Builder {
        private var manufacturer = "Ford"
        private var cylinders: Int = 0
        fun withManufacturer(manufacturer: String): Builder {
            this.manufacturer = manufacturer
            return this
        }

        fun withCylinders(cylinders: Int): Builder {
            this.cylinders = cylinders
            return this
        }

        fun build(crankShaft: CrankShaft.Builder, sparkPlug: SparkPlug.Builder): EngineImpl {
            return EngineImpl(manufacturer, cylinders, crankShaft.build(), sparkPlug.build())
        }
    }

    companion object {
        fun builder(): Builder {
            return Builder()
        }
    }
}
// end::class[]

File: test-suite-python/src/test/python/micronaut/docs/config/builder/EngineImpl.py
from .CrankShaft import CrankShaft
from .CrankShaft import CrankShaftBuilder
from .Engine import Engine
from .SparkPlug import SparkPlug
from .SparkPlug import SparkPlugBuilder


# tag::class[]
class EngineImpl(Engine):
    def __init__(
        self,
        manufacturer: str,
        cylinders: int,
        crank_shaft: CrankShaft,
        spark_plug: SparkPlug,
    ):
        self.manufacturer = manufacturer
        self.cylinders = cylinders
        self.crank_shaft = crank_shaft
        self.spark_plug = spark_plug

    def get_cylinders(self) -> int:
        return self.cylinders

    def start(self) -> str:
        rod_length = self.crank_shaft.rod_length
        if rod_length is None:
            rod_length = 6.0
        return (
            f"{self.manufacturer} Engine Starting V{self.cylinders} "
            f"[rodLength={rod_length}, sparkPlug={self.spark_plug}]"
        )

    @staticmethod
    def builder() -> "EngineImplBuilder":
        return EngineImplBuilder()
# end::class[]


class EngineImplBuilder:
    manufacturer: str = "Ford"
    cylinders: int = 0

    def withManufacturer(self, manufacturer: str) -> "EngineImplBuilder":
        self.manufacturer = manufacturer
        return self

    def withCylinders(self, cylinders: int) -> "EngineImplBuilder":
        self.cylinders = cylinders
        return self

    def build(self, crank_shaft: CrankShaftBuilder, spark_plug: SparkPlugBuilder) -> EngineImpl:
        return EngineImpl(
            self.manufacturer,
            self.cylinders,
            crank_shaft.build(),
            spark_plug.build(),
        )


Builder = EngineImplBuilder
EngineImpl.Builder = EngineImplBuilder
